File: .local/lib/copilot_workflows/memory.py
"""Durable cross-run memory (`wf.memory`): a text file shared across `cwf loop` ticks."""
from __future__ import annotations

import os
import threading
from collections.abc import Callable

from ._util import noop


class Memory:
    """A durable text file shared across runs and loop ticks; thread-safe, no-op when unset."""

    def __init__(self, path: str | None = None, *, read_only: bool = False,
                 logger: Callable[..., None] | None = None):
        self.path = os.path.abspath(os.path.expanduser(path)) if path else None
        self._read_only = read_only
        self._log = logger or noop
        self._lock = threading.Lock()

    @property
    def enabled(self) -> bool:
        """True when a memory file is configured (``--memory PATH`` was passed)."""
        return self.path is not None

    def __bool__(self) -> bool:
        return self.enabled

    def read(self) -> str:
        """Return the file's full contents, or ``""`` if memory is unset or absent."""
        if not self.path:
            return ""
        with self._lock:
            try:
                with open(self.path, encoding="utf-8") as fh:
                    return fh.read()
            except FileNotFoundError:
                return ""
            except OSError as e:
                self._log(f"  ! memory read failed: {e}")
                return ""

    def write(self, text: str) -> None:
        """Overwrite memory with ``text``. No-op when unset or in dry-run."""
        self._put(str(text), append=False)

    def append(self, text: str) -> None:
        """Append ``text`` (newline-terminated) to memory. No-op when unset or in dry-run."""
        self._put(str(text), append=True)

    def clear(self) -> None:
        """Truncate memory to empty. No-op when unset or in dry-run."""
        self._put("", append=False)

    def _put(self, text: str, *, append: bool) -> None:
        if not self.path:
            return
        if self._read_only:
            verb = "append" if append else "write"
            self._log(f"  memory: [dry-run] skipped {verb} ({len(text)} chars)")
            return
        with self._lock:
            try:
                os.makedirs(os.path.dirname(self.path), exist_ok=True)
                with open(self.path, "a" if append else "w", encoding="utf-8") as fh:
                    fh.write(text)
                    if append and text and not text.endswith("\n"):
                        fh.write("\n")  # keep each note on its own line
                    fh.flush()
                    os.fsync(fh.fileno())  # durable across a tick crash, like checkpoints
            except OSError as e:
                self._log(f"  ! memory write failed: {e}")

File: .copilot/extensions/conveyor/memory.mjs
/**
 * @module memory
 *
 * Durable cross-run text memory exposed to a harness as `memory`. Backed by a single file so it can
 * be shared across recurring ticks. No-ops (reads return "") when no path is configured. Under
 * `dryRun` it is read-only: writes are logged as skipped and never mutate the file.
 */
import { readFileSync, mkdirSync, openSync, writeSync, fsyncSync, closeSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { homedir } from "node:os";

import { atomicWriteFile } from "./persistence.mjs";

/** Expand a bare `~` or a leading `~/`. @param {string} p */
const expandHome = (p) => (p === "~" ? homedir() : p.startsWith("~/") ? resolve(homedir(), p.slice(2)) : p);

/** A durable text file shared across runs/ticks; no-op when unset, read-only under dry-run. */
export class Memory {
	/** @type {string | undefined} */
	path;
	#readOnly;
	#log;

	/**
	 * @param {string | null | undefined} path
	 * @param {{ readOnly?: boolean, log?: (m: string) => void }} [opts]
	 */
	constructor(path, { readOnly = false, log = () => {} } = {}) {
		this.path = path ? resolve(expandHome(path)) : undefined;
		this.#readOnly = readOnly;
		this.#log = log;
	}

	/** True when a memory file is configured. */
	get enabled() {
		return this.path !== undefined;
	}

	/** Full file contents, or "" when unset/absent. */
	read() {
		if (!this.path) return "";
		try {
			return readFileSync(this.path, "utf8");
		} catch (e) {
			if (/** @type {NodeJS.ErrnoException} */ (e).code !== "ENOENT") {
				this.#log(`  ! memory read failed: ${e instanceof Error ? e.message : e}`);
			}
			return "";
		}
	}

	/** Overwrite memory. No-op when unset or dry-run. @param {string} text */
	write(text) {
		this.#put(String(text), false);
	}

	/** Append a newline-terminated note. No-op when unset or dry-run. @param {string} text */
	append(text) {
		this.#put(String(text), true);
	}

	/** Truncate memory to empty. No-op when unset or dry-run. */
	clear() {
		this.#put("", false);
	}

	/** @param {string} text @param {boolean} append */
	#put(text, append) {
		if (!this.path) return;
		if (this.#readOnly) {
			this.#log(`  memory: [dry-run] skipped ${append ? "append" : "write"} (${text.length} chars)`);
			return;
		}
		try {
			mkdirSync(dirname(this.path), { recursive: true });
			// Memory is durable cross-run state, so writes are fsync'd like ledger records.
			if (!append) {
				atomicWriteFile(this.path, text);
				return;
			}
			const fd = openSync(this.path, "a");
			try {
				writeSync(fd, text && !text.endsWith("\n") ? text + "\n" : text, null, "utf8");
				fsyncSync(fd);
			} finally {
				closeSync(fd);
			}
		} catch (e) {
			this.#log(`  ! memory write failed: ${e instanceof Error ? e.message : e}`);
		}
	}
}
